CodingPractice/PythonAssignments/tests/cleancode/test_EmailExtract.py

```diff
@@ -0,0 +1,24 @@
+import unittest
+from CodingPractice.PythonAssignments.cleancode.EmailExtract \
+    import EmailDetails, extract_email_address_details, MalformedEmailAddressException
+
+
+
+class EmailExtractTests(unittest.TestCase):
+
+    def test_00_is_id_valid(self):
+        actual = extract_email_address_details( 'basil@5yp.com' )
+        self.assertEqual('basil', actual.id)
+
+    def test_01_can_extract_address(self):
+        expected = EmailDetails('basil', '5yp.com')
+        actual = extract_email_address_details( 'basil@5yp.com' )
+        self.assertEqual(expected, actual)
+
+    def test_02_exception_raised_for_malformed_email_address(self):
+        with self.assertRaises(MalformedEmailAddressException):
+            extract_email_address_details( 'BAD_EMAIL_ADDRESS' )
+
+
```

Comment:
How about a test for 'sufi@' or '@5yp.com' ?
What should extract_email_address_details do in these situations ?

Reply:
How about tests for '@@@' and 'bbb@xyz,com@'

Reply:
I applied Regex, as you will be able to see in my next commit.

Reply:
Interesting choice Sufi. Let's explore that.
Does your regex find the following valid ?
1. abc@123.co.uk
2. 'abc    @   123   ....   abc .... com'
3. a@1.;£$%!&c

:) Maybe you need tests for these